stor/utils.py

```diff
@@ -229,6 +230,85 @@ def is_obs_path(p):
     return is_s3_path(p) or is_swift_path(p)
 
 
+def is_writeable(path):
+    """
+    Determine whether we have permission to write to path.
+
+    Behavior of this method is slightly different for different storage types when the
+    directory doesn't exist:
+    1. For local file systems, this function will return True if the target directory can
+       be created and a file written to it.
+    2. For AWS S3, this function will return True only if the target bucket is already
+       present and we have write access to the bucket.
+    3. For Swift, this function will return True, only if the target tenant is already
+       present and we have write access to the tenant and container. The container doesn't
+       have to be present.
+
+    This is function is useful, because `stor.stat()` will succeed if we have read-only
+    permissions to `path`, but the eventual attempt to upload will fail.
+
+    Secondly, `path` might not exist yet. If the intent of the caller is to create it, ,
+    stor.stat() will fail, however the eventual upload attempt would succeed.
+    """
+    from stor import basename
+    from stor import join
+    from stor import Path
+    from stor import remove
+    from stor import rmtree
+    from stor.swift import SwiftPath
+    from stor.swift import UnauthorizedError
+
+    path = Path(path)
+
+    # We want this function to behave as a no-op with regards to the underlying directory
+    # structure. Therefore we need to remove any directories created by this function that
+    # were not present when it was called. The `base_path_existed` defined below will
+    # store whether the directory that we're checking existed when calling this function,
+    # so that we know if it should be removed at the end.
+    if is_filesystem_path(path):
+        # Need to remove the path, once the test is completed.
+        base_path = path
+        base_path_existed = path.exists()
+    elif is_s3_path(path):
+        # No need to remove anything in S3, since there's no actual hierarchical structure
+        # there. The bucket has to exist before this function is called.
+        base_path = None
+        base_path_existed = None
+    elif is_swift_path(path):
+        # Need to remove container after the test if it didn't exist before this function
+        # was called.
+        base_path = Path('{}{}/{}'.format(
+            SwiftPath.drive,
+            path.tenant,
+            path.container
+        ))
+        base_path_existed = base_path.exists()
+    else:  # pragma: no cover
+        raise ValueError('Path type not supported: {}'.format(path))
+
+    with tempfile.NamedTemporaryFile() as tmpfile:
+        try:
+            if is_filesystem_path(path):
+                make_dest_dir(path)
+
+            # Attempt to create a file in the `path`.
+            copy(tmpfile.name, path)
+            answer = True
+        except (UnauthorizedError, IOError, OSError, exceptions.FailedUploadError):
```

Comment:
`copy` calls `SwiftPath.upload`, which already is decorated with `_swift_retry` decorator that implements exponential backoff.

>though we might want to consider allowing user to say "I want a fast is_writeable here!"

bff3dd6c97033fe8ed4ca38b9aca5a4d72f66ea8